R/checkPythonDeps.R

```diff
@@ -0,0 +1,350 @@
+        return(invisible(FALSE))
+      }
+    )
+  } else {
+    return(invisible(TRUE))
+  }
+}
+
+#' Check Python Dependencies and Versions
+#'
+#' Compares requirements given in a pip-style file with the installed Python packages. Optionally check if versions are
+#' met.
+#'
+#' @param requirementsFile Path to a file containing the Python requirements. pip-style requirements are expected
+#' @param installed Python packages installed in the environment. Defaults to reticulate::py_list_packages()
+#' @param action Action if a dependency or version is missing or mismatched. Either "stop", "warn", "note", or "pass"
+#' @param strict Logical indicating whether to check for strict version matching. Default is TRUE
+#' @return Vector of missing packages.
+#' @export
+checkPythonRequirements <- function(requirementsFile, installed = py_list_packages(), action = "stop", strict = TRUE) {
+  # Sanity checks for file ...
+  if (!file.exists(requirementsFile)) {
+    stop("Requirements file '", requirementsFile, "' does not exist.")
+  }
+  # ... and action
+  if (!action %in% c("stop", "warn", "note", "pass")) {
+    stop("Invalid action '", action, "'. Must be one of 'stop', 'warn', 'note' or 'pass'.")
+  }
+  # Read the requirements file
+  requirements <- readLines(requirementsFile)
+  # Filter out empty or whitespace-only strings and sort alphabetically
+  requirements <- sort(requirements[!grepl("^\\s*$", requirements)])
+  requiredDependencies <- map(requirements, extractPythonDependency, style = "pip")
+  requiredPackages <- map_chr(requiredDependencies, "name")
+  # Now check if the required packages are installed ...
+  missing <- !(requiredPackages %in% installed$package)
+  present <- installed$package %in% requiredPackages
+  # ... and in strict mode if they have the correct versions
+  if (strict && all(!missing)) {
+    # Use purrr's list comprehension to see if the installed package versions agree with the requirements. pmap applies
+    # the comparison function to the list of lists element-wise. The result is a logical vector, one for each package,
+    # indicating whether the version fullfills the requirement
+    nomatch <- unlist(pmap(
+      # We need to consider three lists: The comparison operators, the installed versions and the required versions
+      list(
+        map(requiredDependencies, "operator"), # Extract comparison operators from depencies
+        # From all dependencies that are present in the Python environment, extract the version strings and convert them
+        map(installed$version[present], extractPythonVersion),
+        map(requiredDependencies, "version") # Extract required versions from deps
+      ),
+      comparePythonVersions
+    ))
+  } else {
+    # If we're not in strict mode, skip the version comparison and assume all versions are correct
+    nomatch <- rep(TRUE, length(requiredPackages))
+  }
+  # Error handling when dependencies are missing or versions mismatch
+  switch(action,
```

Comment:
This is a lot of code duplication. What about something like this:
```
f <- switch(action,
  stop = stop,
  warn = warning,
  note = message,
  pass = function(...) TRUE
)
```
And then you only need the block once and instead of e.g. `stop` you call `f`?

Reply:
I like the idea, however if the stop/warn messaging is moved to a dedicated function (let's call it `notifyUser`), the origin of the error is obfuscated, e.g.:

```
anotherFunction <- function() {
  [..do things, error occurs..]
  notifyUser("stop", "foobar")
}
```

yields `Error in notifyUser("stop", "error message") : error message` instead of `Error in anotherFunction [...]`.

I consolidated the error message creation now

Reply:
What a shame, and it seems stop does not have a way to include the previous call on the call stack in the error message instead :/